### PR TITLE
Added ManageIQ Status with conditions, versions and ingress endpoint info

### PR DESCRIPTION
**Pull request by sudhir-kelkar**
1. This PR adds support for ManageIQ CR status.
With this status implementation we are adding status conditions, ingress endpoint info and operator/operand version to the CR status.

Fixes `#https://github.ibm.com/katamari/dev-issue-tracking/issues/23259`

### CODE REVIEW

File: manageiq-operator/controllers/manageiq_controller.go [hunk 1]
```diff
@@ -166,6 +173,181 @@ func (r *ManageIQReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
+					object.Spec.Rules != nil {
+					endpointInfo := &miqv1alpha1.Endpoint{}
+					if len(object.Spec.TLS[0].SecretName) != 0 {
+						if objectSecret := FindSecret(cr, r.Client, object.Spec.TLS[0].SecretName); object != nil {
+							endpointInfo.Name = ingressName
+							endpointInfo.Type = "UI"
+							endpointInfo.Scope = "External"
+							for k := range objectSecret.Data {
+								if k == "cs.crt" {
+									endpointInfo.CASecret.Key = "cs.crt"
```
Comment (bdunne) [inline manageiq-operator/controllers/manageiq_controller.go:251]: ca.crt

**Review by bdunne — COMMENTED** [review]
Apparently My comments from last week were not submitted... trying again

File: manageiq-operator/controllers/manageiq_controller.go [hunk 2]
```diff
@@ -166,6 +243,91 @@ func (r *ManageIQReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
+		},
+		{
+			Name: "operand", Version: operandVersion,
+		},
+	}
+	return nil
+}
+
+func (r *ManageIQReconciler) reportStatusCondition(miqInstance *miqv1alpha1.ManageIQ, statusMessage string,
+	reason string, conditionStatus metav1.ConditionStatus, statusType string) {
```
Comment (bdunne) [inline manageiq-operator/controllers/manageiq_controller.go:306]: Can this remain on one line?

File: manageiq-operator/controllers/manageiq_controller.go [hunk 3]
```diff
@@ -147,10 +151,83 @@ func (r *ManageIQReconciler) Reconcile(ctx context.Context, request ctrl.Request
...
+				conditionMessage := fmt.Sprintf("[%s] %s", typeReplicaFailure.Type, typeReplicaFailure.Message)
+				r.reportStatusCondition(miqInstance, conditionMessage, typeReplicaFailure.Reason, metav1.ConditionStatus(typeReplicaFailure.Status), deploymentName)
+			} else if typeProgressingCondition := FindDeploymentStatusCondition(deploymentStatusConditions, appsv1.DeploymentProgressing); typeProgressingCondition != nil {
+				conditionMessage := fmt.Sprintf("[%s] %s", typeProgressingCondition.Type, typeProgressingCondition.Message)
+				r.reportStatusCondition(miqInstance, conditionMessage, typeProgressingCondition.Reason, metav1.ConditionStatus(typeProgressingCondition.Status), deploymentName)
+			} else if typeAvailableCondition := FindDeploymentStatusCondition(deploymentStatusConditions, appsv1.DeploymentAvailable); typeAvailableCondition != nil {
+				conditionMessage := fmt.Sprintf("[%s] %s", typeAvailableCondition.Type, typeAvailableCondition.Message)
+				r.reportStatusCondition(miqInstance, conditionMessage, typeAvailableCondition.Reason, metav1.ConditionStatus(typeAvailableCondition.Status), deploymentName)
+			}
```
Comment (bdunne) [inline manageiq-operator/controllers/manageiq_controller.go:188]: Is it possible to do something like this?
```suggestion
for i, conditionType := range [appsv1.DeploymentReplicaFailure, appsv1.DeploymentProgressing, appsv1.DeploymentAvailable] {
  for i, condition := range conditions {
    if condition.Type == conditionType {
      r.reportStatusCondition(miqInstance, fmt.Sprintf("[%s] %s", condition.Type, condition.Message), condition.Reason, metav1.ConditionStatus(condition.Status), deploymentName)
    }
  }
}
```

**Comment by bdunne** [comment]
@Fryguy Can you also take a look at this?

**Comment by Fryguy** [comment]
Overall LGTM, though I admit I'm not a go person.  I agree with @bdunne's comments otherwise.

File: manageiq-operator/controllers/manageiq_controller.go (hunk 2 above)
Comment (sudhir-kelkar) [reply manageiq-operator/controllers/manageiq_controller.go]: done

File: manageiq-operator/controllers/manageiq_controller.go (hunk 3 above)
Comment (sudhir-kelkar) [reply manageiq-operator/controllers/manageiq_controller.go]: Had a call with Brandon. Also discussed for loop approach.
After discussion we finalize on if elseif approach

File: manageiq-operator/controllers/manageiq_controller.go [hunk 4]
```diff
@@ -18,23 +18,28 @@ package controllers
 
 import (
 	"context"
+	"fmt"
+
+	"os"
```
Comment (bdunne) [inline manageiq-operator/controllers/manageiq_controller.go:23]:
```suggestion
	"fmt"
	"os"
```

File: manageiq-operator/controllers/manageiq_controller.go [hunk 5]
```diff
@@ -18,23 +18,28 @@ package controllers
...
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/types"
-	"os"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
 	"sigs.k8s.io/controller-runtime/pkg/log"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 
+	"github.com/ManageIQ/manageiq-pods/manageiq-operator/api/v1alpha1"
```
Comment (bdunne) [inline manageiq-operator/controllers/manageiq_controller.go:37]:
```suggestion
```

File: manageiq-operator/controllers/manageiq_controller.go [hunk 6]
```diff
@@ -166,6 +247,90 @@ func (r *ManageIQReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
+			return &conditions[i]
+		}
+	}
+	return nil
+}
+
+func (r *ManageIQReconciler) reportOperatorVersions(miqInstance *miqv1alpha1.ManageIQ) error {
+	operatorVersion := miqInstance.Spec.AppAnnotations["productVersion"]
+	operandVersion := miqInstance.APIVersion
+	miqInstance.Status.Versions = []v1alpha1.Version{
```
Comment (bdunne) [inline manageiq-operator/controllers/manageiq_controller.go:298]:
```suggestion
	miqInstance.Status.Versions = []miqv1alpha1.Version{
```

File: manageiq-operator/controllers/manageiq_controller.go [hunk 7]
```diff
@@ -166,6 +247,90 @@ func (r *ManageIQReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
+func (r *ManageIQReconciler) reportStatusCondition(miqInstance *miqv1alpha1.ManageIQ, statusMessage string, reason string, conditionStatus metav1.ConditionStatus, statusType string) {
+	apimeta.SetStatusCondition(&miqInstance.Status.Conditions, metav1.Condition{
+		Type:    statusType,
+		Status:  conditionStatus,
+		Reason:  reason,
+		Message: statusMessage,
+	})
+}
+
+func (r *ManageIQReconciler) reportEndpointInfo(miqInstance *miqv1alpha1.ManageIQ, endPointDetails v1alpha1.Endpoint) {
```
Comment (bdunne) [inline manageiq-operator/controllers/manageiq_controller.go:318]:
```suggestion
func (r *ManageIQReconciler) reportEndpointInfo(miqInstance *miqv1alpha1.ManageIQ, endPointDetails miqv1alpha1.Endpoint) {
```

File: manageiq-operator/api/v1alpha1/zz_generated.deepcopy.go [hunk 8]
```diff
@@ -22,16 +22,33 @@ limitations under the License.
 package v1alpha1
 
 import (
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
```
Comment (bdunne) [inline manageiq-operator/api/v1alpha1/zz_generated.deepcopy.go:25]:
```suggestion
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```

File: manageiq-operator/api/v1alpha1/zz_generated.deepcopy.go [hunk 9]
```diff
@@ -134,6 +151,23 @@ func (in *ManageIQSpec) DeepCopy() *ManageIQSpec {
...
+		copy(*out, *in)
+	}
+	if in.Endpoints != nil {
+		in, out := &in.Endpoints, &out.Endpoints
+		*out = make([]Endpoint, len(*in))
+		copy(*out, *in)
+	}
+	if in.Conditions != nil {
+		in, out := &in.Conditions, &out.Conditions
+		*out = make([]v1.Condition, len(*in))
```
Comment (bdunne) [inline manageiq-operator/api/v1alpha1/zz_generated.deepcopy.go:166]:
```suggestion
		*out = make([]metav1.Condition, len(*in))
```

File: manageiq-operator/controllers/manageiq_controller.go [hunk 10]
```diff
@@ -166,6 +247,90 @@ func (r *ManageIQReconciler) SetupWithManager(mgr ctrl.Manager) error {
 
 var logger = log.Log.WithName("controller_manageiq")
 
+func FindDeployment(cr *miqv1alpha1.ManageIQ, client client.Client, name string) *appsv1.Deployment {
```
Comment (bdunne) [inline manageiq-operator/controllers/manageiq_controller.go:250]: In a followup, do you think we should lower case these methods? I think they can be private.

File: manageiq-operator/api/v1alpha1/zz_generated.deepcopy.go (hunk 8 above)
Comment (bdunne) [reply manageiq-operator/api/v1alpha1/zz_generated.deepcopy.go:25]: Nevermind, the generator overwrites this.

File: manageiq-operator/api/v1alpha1/zz_generated.deepcopy.go (hunk 9 above)
Comment (bdunne) [reply manageiq-operator/api/v1alpha1/zz_generated.deepcopy.go:166]: Nevermind, the generator overwrites this.

File: manageiq-operator/controllers/manageiq_controller.go (hunk 4 above)
Comment (bdunne) [reply manageiq-operator/controllers/manageiq_controller.go]: I pushed a commit for this

File: manageiq-operator/controllers/manageiq_controller.go (hunk 5 above)
Comment (bdunne) [reply manageiq-operator/controllers/manageiq_controller.go]: I pushed a commit for this

File: manageiq-operator/controllers/manageiq_controller.go (hunk 6 above)
Comment (bdunne) [reply manageiq-operator/controllers/manageiq_controller.go]: I pushed a commit for this

File: manageiq-operator/controllers/manageiq_controller.go (hunk 7 above)
Comment (bdunne) [reply manageiq-operator/controllers/manageiq_controller.go]: I pushed a commit for this

**Pull request merged**